group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupRegexManager.java

```diff
@@ -0,0 +1,388 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.coordinator.group;
+
+import org.apache.kafka.common.errors.InvalidRegularExpression;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.coordinator.common.runtime.CoordinatorRecord;
+import org.apache.kafka.coordinator.common.runtime.CoordinatorResult;
+import org.apache.kafka.coordinator.common.runtime.CoordinatorTimer;
+import org.apache.kafka.coordinator.group.modern.consumer.ConsumerGroupRegex.RegexKey;
+import org.apache.kafka.coordinator.group.modern.consumer.ConsumerGroupRegex.Resolution;
+import org.apache.kafka.image.MetadataImage;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+
+import com.google.re2j.Pattern;
+import com.google.re2j.PatternSyntaxException;
+
+import org.slf4j.Logger;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * The GroupRegexManager maintains regular expressions used in pattern-based subscription for all consumer groups,
+ * resolved using RE2J. It's responsible for:
+ * 1) Validating regular expressions.
+ * 2) Providing the resolution of a regular expression if available.
+ * 3) Evaluating regular expressions to find the matching topics (performed in a separate thread)
+ * 4) Re-evaluating regular expressions periodically to keep them up-to-date as topics are created/deleted.
+ */
+public class GroupRegexManager {
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+    /**
+     * Queue containing regular expressions that need to be evaluated.
+     */
+    private final UniqueBlockingQueue<RegexKey> evalQueue;
+
+    /**
+     * The classic and consumer groups keyed by their name.
+     */
+    private final TimelineHashMap<RegexKey, Resolution> resolvedRegexes;
+
+    /**
+     * The metadata image.
+     */
+    private MetadataImage metadataImage;
+
+    /**
+     * The system timer.
+     */
+    private final CoordinatorTimer<Void, CoordinatorRecord> timer;
+
+    /**
+     * Indicates if the manager should run the asynchronous eval of requested regular expressions. This is expected
+     * to be true from the creation of the manager until there is a call to {@link #onUnloaded()}.
+     */
+    private boolean runAsyncEval;
+
+    /**
+     * Executor to run the thread for evaluating regular expressions.
+     */
+    private final ExecutorService executorService;
```

Comment:
In this design, we use an executor per shard. By default, we use 50 partitions. If they are all hosted on the same coordinator, we would have 50 threads. I wonder if we could rather have a shared executor used by all the shards. Have you considered it?